### ISSUE TITLE
ephemeral test suites encounter flag redefine error

### ISSUE THREAD

**Issue by saikat-royc** [kubernetes/kubernetes#90868]
**What happened**: 
On running kubetest with external storage driver option with driver name "pd.csi.storage.gke.io", encountered error of flag redefinition.
The issue happens because:
As part of this patch https://github.com/kubernetes/kubernetes/commit/7d2ad805a975b5a68e84a650aba92895441170b7, a flag is defined for number of inline volumes.
The driver name is used to construct the flag name.

Thus, when running tests on an externally provided driver (in this instance the Persistent Disk CSI driver) with driver name "pd.csi.storage.gke.io", a redefinition error happens since the ephemeral test suite code path is executed once for https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go 
and once for 
External PD CSI driver.
[example run command: 
kubetest --test --check-version-skew=false --test_args="--ginkgo.focus=External.*Storage.*snapshot --storage.testdriver=/path/to/gcp-compute-persistent-disk-csi-driver/test/k8s-integration/config/test-config.yaml"

Test config looks something like this:
{
StorageClass:
  FromFile: ...
SnapshotClass:
  FromFile: ...
DriverInfo:
  Name: pd.csi.storage.gke.io
...
}
]

**What you expected to happen**:
Test should run and succeed.

**How to reproduce it (as minimally and precisely as possible)**:
Use an test-config.yaml, with driver name as "pd.csi.storage.gke.io", and run kubetest with focus as "External.Storage"

kubetest --test --check-version-skew=false --test_args="--ginkgo.focus=External.*Storage.*snapshot --storage.testdriver=/path/to/gcp-compute-persistent-disk-csi-driver/test/k8s-integration/config/test-config.yaml"

**Anything else we need to know?**:
Possible fixes options:
1. In, external.go (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/external/external.go#L343), when using a FromFile or FromExistingClassName option to pass in SnapshotClass information, pickup the provisioner name from the class definition's "driver" field.

2. Remove the flag which has a dependency on driver names. (need to understand the impact of removing the command line flag)

**Environment**:
- Kubernetes version (use `kubectl version`): latest master
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by msau42** [comment]
/assign @pohly 
Do you know why this parameter is set as a e2e.test flag and not a part of the [DriverInfo](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L160)?

**Comment by pohly** [comment]
> Do you know why this parameter is set as a e2e.test flag and not a part of the DriverInfo?

To me, `DriverInfo` is something that defines what the driver can do. This flag is about what the test does and not related to driver capabilities.

Making it depend on the driver name was probably unnecessary. A single flag for all incarnations of that test probably would have been sufficient and thus better ("keep it simple, stupid!" - the stupid being me here...).

It also shouldn't have used the global flag set.

In other words:
```
import e2econfig "k8s.io/kubernetes/test/e2e/framework/config"

var numInlineVolumes = e2econfig.Flags.Int("storage.ephemeral.numInlineVolumes", 2, "number of ephemeral inline volumes per pod")
```

**Comment by msau42** [comment]
/reopen

I don't really want to get into the habit of defining per test case flags in e2e.test. There are thousands of test cases and it would be unwieldy to discover all the flags available and understand how to use them.

What factors would influence how many volumes you want the test to create? Can we just hardcode a value in the test case?

**Comment by pohly** [comment]
> I don't really want to get into the habit of defining per test case flags in e2e.test. There are thousands of test cases and it would be unwieldy to discover all the flags available and understand how to use them.

The defaults must be usable so that users do not need to understand them. The flags can be discovered with `go test ./test/e2e -args -help` - not ideal, but at least they show up.

> Can we just hardcode a value in the test case?

That's effectively what the default value for the flag does.

> What factors would influence how many volumes you want the test to create? 

This is a way to to run a custom stress test: the default is so low, it just tests that the feature works. By increasing it as high as the underlying CSI driver and storage system supports, one can run this (and potentially other tests) as part of stress testing.

Not sure whether it's worth supporting this. It seemed worthwhile at the time, and better than hard-coding just one value.

**Comment by pohly** [comment]
The bigger issue is: how do users of the test suite actually learn what tests they run and what those tests do? I suspect the answer to that is simply "read the source", in which case they'll also find the knobs that can be used to tweak testing.

**Comment by msau42** [comment]
Since increasing the parameter depends on the csi driver/storage system, then I would recommending splitting this up into 2 test cases: 1) test basic functionality 2) test stress scenario, and add the parameter to the driver test config similarly to other stress tests: https://github.com/kubernetes/kubernetes/blob/9b44fa4a07b14b3dbc644f80e087d516677632f0/test/e2e/storage/testsuites/testdriver.go#L196

**Comment by msau42** [comment]
At least for the storage tests, because we are intending that they be run externally, having all the tunables in one place I think makes it easier to discover.

**Comment by mauriciopoppe** [comment]
@pohly I saw that the flag was removed in https://github.com/kubernetes/kubernetes/commit/2468a24b7a732fa492027a159ee15b6d31bf0577#diff-20517805986874f69e3254bed93d59996d5a6fd571a70ab8120736ded5aafa24, @msau42 I think we can close this one if https://github.com/kubernetes/kubernetes/issues/90868#issuecomment-626964330 doesn't apply anymore

**Comment by mauriciopoppe** [comment]
/close
The flag was removed in 2468a24#diff-20517805986874f69e3254bed93d59996d5a6fd571a70ab8120736ded5aafa24
